### PR TITLE
err objects could not be compared directly

### PR DESCRIPTION
**Pull request by zhangxiaoyu-zidif**
**What this PR does / why we need it**:
err objects could not be compared directly

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/daemon/daemon_controller.go [hunk 1]
```diff
@@ -1239,7 +1239,7 @@ func (dsc *DaemonSetsController) nodeShouldRunDaemonPod(node *v1.Node, ds *exten
 	wantToRun, shouldSchedule, shouldContinueRunning = true, true, true
 	// If the daemon set specifies a node name, check that it matches with node.Name.
 	if !(ds.Spec.Template.Spec.NodeName == "" || ds.Spec.Template.Spec.NodeName == node.Name) {
-		return false, false, false, nil
+		return false, false, false, fmt.Errorf("specified node name does not match node.Name")
```
Comment (dixudx) [inline pkg/controller/daemon/daemon_controller.go]: How about "Spec.NodeName does not match node's name"?

Comment (zhangxiaoyu-zidif) [reply pkg/controller/daemon/daemon_controller.go]: sounds good to me.

**Comment by zhangxiaoyu-zidif** [comment]
@mikedanese PTAL :)

**Comment by zhangxiaoyu-zidif** [comment]
@wojtek-t 
could you take a look. =) thanks.

**Comment by zhangxiaoyu-zidif** [comment]
@mikedanese
could you take a look please

File: pkg/controller/daemon/daemon_controller.go [hunk 2]
```diff
@@ -1239,7 +1239,7 @@ func (dsc *DaemonSetsController) nodeShouldRunDaemonPod(node *v1.Node, ds *exten
 	wantToRun, shouldSchedule, shouldContinueRunning = true, true, true
 	// If the daemon set specifies a node name, check that it matches with node.Name.
 	if !(ds.Spec.Template.Spec.NodeName == "" || ds.Spec.Template.Spec.NodeName == node.Name) {
-		return false, false, false, nil
+		return false, false, false, fmt.Errorf("Spec.NodeName does not match node's name")
```
Comment (mikedanese) [inline pkg/controller/daemon/daemon_controller.go]: This just means that the pod shouldn't run on that node. That's not an error. This breaks all usages of nodeShouldRunDaemonPod in this controller.

**Comment by mikedanese** [comment]
Look at where nodeShouldRunDaemonPod is used in daemon_controller.go. This change would break the behavior of all those call sites.

**Comment by zhangxiaoyu-zidif** [comment]
@mikedanese 
Thank you sir, my bad, I consider it rashly. 
There is a nit here. two error object can not be compared... so we should change it to string.

File: pkg/controller/daemon/daemon_controller_test.go [hunk 3]
```diff
@@ -1830,7 +1830,7 @@ func TestNodeShouldRunDaemonPod(t *testing.T) {
 			if shouldContinueRunning != c.shouldContinueRunning {
 				t.Errorf("[%v] strategy: %v, predicateName: %v expected shouldContinueRunning: %v, got: %v", i, c.ds.Spec.UpdateStrategy.Type, c.predicateName, c.shouldContinueRunning, shouldContinueRunning)
 			}
-			if err != c.err {
+			if err.Error() != c.err.Error() {
```
Comment (zhangxiaoyu-zidif) [inline pkg/controller/daemon/daemon_controller_test.go]: @mikedanese 
we could compare two error object directly. They could not be equal, even they have the same error messages.

**Comment by deads2k** [comment]
> /assign @deads2k

/unassign

Looks like @mikedanese was already looking at this one (had prior comments).

**Comment by zhangxiaoyu-zidif** [comment]
ping @mikedanese

**Pull request closed without merging**